AI blockade avoidance

Dilvish-fo opened the pull request:
This allows the AI to use the information from Empire::UnrestrictedLaneTravel() to have its non-warship fleets plot routes around blockaded starlanes (in the existing code, they generally just hold off from trying to pass through any system with more enemy forces than one's own, and wait for reinforcements to come).

For testing, there is a _DEBUG_CHAT variable that can be set to get chat messages when the AI has rerouted around a danger.

In the discussion in #1761 there was some mention of possibly having the AI pass in a list of fully prohibited systems to be routed around, and that could still be a useful augmentation after this PR, particularly useful in situations where the AI knew it was planning on withdrawing forces from a contested system, for example.

Dilvish-fo committed:
Add AI decorator cache_by_session_with_turnwise_update
  function decorator that caches a function value during session,
  updated each turn.

Morlic-fo reviewed (changes requested):
Some style issues but generally looks fine to me.

Dilvish-fo committed:
Have the AI avoid blockaded starlane travel for passive fleets
 - passive fleets can still enter a blockaded final destination system
 - aggressive fleets will not try to route around blockades
 - makes find_path_with_resupply more generic and renames it
 - adds new find_path_with_resupply that specifies various details and then
   calls the generic version
 - determines AI pathfinding policy by fleet mission type

 - has a scouting option that allows to plot an otherwise safe route through
   a system which has never been partially visible before (and hence counts
   as fully blockaded)

Note that this is specifically regarding routing-- as it stands, in the
situation where there are hostile forces in a system, but the routed path
is not blockaded for the AI, this routing code could allow that path, but
other AI code (in OrderMove) may in some cases currently prevent a fleet
from actually entering the contested system.  Such a decision might still be warranted if the starlane of approach required more than one turn to travel.
